File: apps/backend/src/main/kotlin/com/lyra/app/forms/application/find/FormFinder.kt
package com.lyra.app.forms.application.find

import com.lyra.app.forms.domain.Form
import com.lyra.app.forms.domain.FormFinderRepository
import com.lyra.app.forms.domain.FormId
import com.lyra.app.forms.domain.exception.FormNotFoundException
import com.lyra.app.organization.domain.OrganizationId
import com.lyra.common.domain.Service
import org.slf4j.LoggerFactory

/**
 * This is a service class that handles the finding of forms.
 * It uses a [FormFinderRepository] to find a form by its ID.
 *
 * @property finder [FormFinderRepository] The repository to use for finding forms.
 */
@Service
class FormFinder(private val finder: FormFinderRepository) {
    /**
     * This function is used to find a form by its ID.
     *
     * @param organizationId The organization ID of the form to find.
     * @param formId The ID of the form to find.
     * @return The form if found, or throws a [FormNotFoundException] if not found.
     */
    suspend fun find(organizationId: OrganizationId, formId: FormId): Form? {
        log.debug("Finding form with ids: {}, {}", organizationId, formId)
        return finder.findByFormIdAndOrganizationId(organizationId = organizationId, formId = formId)
    }

    companion object {
        private val log = LoggerFactory.getLogger(FormFinder::class.java)
    }
}

File: apps/frontend/packages/api-services/src/forms/application/find/FormFinder.ts
import FormNotFoundError from '~/forms/domain/FormNotFoundError.ts';
import { inject, injectable } from 'inversify';
import type FormFinderRepository from '~/forms/domain/FormFinderRepository.ts';
import { FORM_FINDER_REPOSITORY_PROVIDER } from '~/di/forms/forms.module.types.ts';
import type FormResponse from '~/forms/domain/FormResponse.ts';

@injectable()
export default class FormFinder {
	constructor(
		@inject(FORM_FINDER_REPOSITORY_PROVIDER) private readonly repository: FormFinderRepository
	) {
    /* ... */
  }

	async find(formId: string): Promise<FormResponse> {
		const form = await this.repository.find(formId);

		if (!form) {
			throw new FormNotFoundError(formId);
		}

		return form;
	}
}
